### PR TITLE
add src files

### PR DESCRIPTION
**Pull request by rolandpo**
add initial src files

### CODE REVIEW

File: src/handle_finalize.c [hunk 1]
```diff
@@ -1,21 +1,39 @@
...
-        msg->numScreens += 1;
-    }
-
-    // EDIT THIS: set `tokenLookup1` (and maybe `tokenLookup2`) to point to
-    // token addresses you will info for (such as decimals, ticker...).
-    msg->tokenLookup1 = context->token_received;
-
-    msg->result = ETH_PLUGIN_RESULT_OK;
-}
+void handle_finalize(void *parameters) {
```
Comment (rafaelugolini) [inline src/handle_finalize.c]: why did you change the type here?

Comment (rolandpo) [reply src/handle_finalize.c]: that's seems to be a new change from 4 days ago. I can try and switch that out and see what else needs changing to make it work, if it breaks anything etc

File: Makefile [hunk 2]
```diff
@@ -24,9 +24,8 @@ include $(BOLOS_SDK)/Makefile.defines
 ########################################
 #        Mandatory configuration       #
 ########################################
-# EDIT THIS: Application name
-# Will be displayed on screen by the Ethereum app
-APPNAME = "PluginBoilerplate"
+# Application name
+APPNAME = "origindefi"
```
Comment (rafaelugolini) [inline Makefile]: can you change the app name to `origin`?

File: README.md [hunk 3]
```diff
@@ -1,40 +1,106 @@
...
 done on a "per contract" basis, meaning a plugin is required for every DApp.
 
-The code has been commented, and special "EDIT THIS" comments indicate where
-developers are expected to adapt the code to their own needs.
+## Prerequisite
+
+Clone the plugin to a new folder.
+
+```shell
+git clone https://github.com/rolandpo/app-plugin-origindefi.git
```
Comment (rafaelugolini) [inline README.md]: can you fix this to match the correct repo?

File: README.md [hunk 4]
```diff
@@ -1,40 +1,106 @@
...
+```
+
+Then in the same folder clone two more repositories, which is the plugin-tools and app-ethereum.
+
+```shell
+git clone --recurse-submodules https://github.com/LedgerHQ/app-ethereum
+sudo docker run --rm -ti -v "$(realpath .):/app" --user $(id -u $USER):$(id -g $USER) ghcr.io/ledgerhq/ledger-app-builder/ledger-app-dev-tools:latest
```
Comment (rafaelugolini) [inline README.md]: why do we need sudo here?

File: README.md [hunk 5]
```diff
@@ -1,40 +1,106 @@
...
 
-It is STRONGLY recommended to follow the
-[plugin guide](https://developers.ledger.com/docs/dapp/embedded-plugin/code-overview/)
-in order to better understand the flow and the context for plugins.
+## Build
+
+The script will build a docker image and attach a console.
+When the docker image is running go to the "app-plugin-origindefi" folder and build the ".elf" files.
+```shell
+cd app-plugin-origindefi/tests       # go to the tests folder in app-plugin-origindefi
```
Comment (rafaelugolini) [inline README.md]: fix all the dir reference from `app-plugin-origindefi`

File: README.md
Comment (rafaelugolini) [inline README.md:1]: I think you just replaced the old README.md file and removed some important documentation pieces/fixes. Can you please recheck that?

File: src/dbg/printf.c
Comment (rafaelugolini) [inline src/dbg/printf.c:1]: do we need this?

File: README.md (hunk 4 above)
Comment (rolandpo) [reply README.md]: this is just copied over from docs: https://developers.ledger.com/docs/dapp/embedded-plugin/testing/#build-the-plugin-and-the-ethereum-app

File: src/handle_finalize.c (hunk 1 above)
Comment (rafaelugolini) [reply src/handle_finalize.c]: I see

File: src/dbg/printf.c
Comment (rolandpo) [reply src/dbg/printf.c:1]: probably not, removed

File: README.md
Comment (rolandpo) [reply README.md:1]: readded the tests at the top

File: README.md (hunk 5 above)
Comment (rolandpo) [reply README.md]: replaced all origin_defi with origin

File: .clang-format [hunk 6]
```diff
@@ -12,7 +12,6 @@ SortIncludes: false
 SpaceAfterCStyleCast: true
 AllowShortCaseLabelsOnASingleLine: false
 AllowAllArgumentsOnNextLine: false
-AllowAllParametersOfDeclarationOnNextLine: false
```
Comment (rafaelugolini) [inline .clang-format:15]: why was this removed?

Comment (rolandpo) [reply .clang-format:15]: it's repeated on a different line, didn't let me format

File: src/handle_init_contract.c [hunk 7]
```diff
@@ -1,51 +1,96 @@
...
-    if (!find_selector(U4BE(msg->selector, 0), SELECTORS, SELECTOR_COUNT, &index)) {
-        PRINTF("Error: selector not found!\n");
+    uint32_t selector = U4BE(msg->selector, 0);
+    if (find_selector(selector, ORIGIN_SELECTORS, NUM_SELECTORS, &context->selectorIndex)) {
         msg->result = ETH_PLUGIN_RESULT_UNAVAILABLE;
         return;
     }
-    context->selectorIndex = index;
-    // check for overflow
-    if ((size_t) context->selectorIndex != index) {
```
Comment (rafaelugolini) [inline src/handle_init_contract.c:34]: Should we keep this overflow check?

File: src/handle_query_contract_ui.c [hunk 8]
```diff
@@ -1,95 +1,325 @@
-#include "plugin.h"
+#include "origin_plugin.h"
 
-// EDIT THIS: You need to adapt / remove the static functions (set_send_ui, set_receive_ui ...) to
-// match what you wish to display.
+static void set_send_ui(ethQueryContractUI_t *msg, origin_parameters_t *context) {
```
Comment (rafaelugolini) [inline src/handle_query_contract_ui.c]: Shouldn't this return `bool` or it doesn't matter?

File: src/handle_init_contract.c (hunk 7 above)
Comment (rolandpo) [reply src/handle_init_contract.c:34]: i can probably add it

File: src/handle_query_contract_ui.c (hunk 8 above)
Comment (rolandpo) [reply src/handle_query_contract_ui.c]: i don't think it matters, but it's based off a previous version of the boilerplate, wouldn't want to change too much unless someone from ledger thinks it's necessary

File: src/handle_init_contract.c (hunk 7 above)
Comment (rafaelugolini) [reply src/handle_init_contract.c:34]: what you did here shouldn't work, you declaring `index` and assigning it to `context-selectorIndex`. 

if you look at the previous code, they pass `index` to `find_selector` as an argument and then they do the check.

File: src/origin_plugin.h [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    } else if (ADDRESS_IS_USDT(context->contract_address_sent)) {
+        context->decimals_sent = USDT_DECIMALS;
+        context->tokens_found |= TOKEN_SENT_FOUND;
+    } else {
+        return false;
+    }
+    return true;
+}
+
+static inline check_token_received(origin_parameters_t *context) {
```
Comment (rafaelugolini) [inline src/origin_plugin.h]: is this suppose to return `bool` or `int`?

File: src/origin_plugin.h [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+_Static_assert(sizeof(origin_parameters_t) <= 5 * 32, "Structure of parameters too big.");
+
+void handle_provide_parameter(void *parameters);
+void handle_query_contract_ui(void *parameters);
+void handle_init_contract(void *parameters);
+void handle_finalize(void *parameters);
+void handle_provide_token(void *parameters);
+void handle_query_contract_id(void *parameters);
+
+static inline check_token_sent(origin_parameters_t *context) {
```
Comment (rafaelugolini) [inline src/origin_plugin.h]: is this suppose to return `bool` or `int`?

File: src/handle_init_contract.c (hunk 7 above)
Comment (rolandpo) [reply src/handle_init_contract.c:34]: yes, that's right of course

File: src/origin_plugin.h (hunk 9 above)
Comment (rolandpo) [reply src/origin_plugin.h]: bool, i added the return type

**Review by rafaelugolini — APPROVED** [review]
